### ISSUE TITLE
ASKING to the community: Is 2-WEEKS RELEASE cycle is fine or we should release LESS FREQUENTLY?

### ISSUE THREAD

**Issue by hikalkan** [abpframework/abp#4692]
We want to get feedback from the community about the release cycle of the ABP Framework

We are releasing a minor/feature version, like 2.3.0, 2.4.0... in every two weeks for a long time. The main motivation behind this decision was to **deliver the value early** and keep our core team motivated and dynamic. Since we are rapidly adding new features and making enhancements, we don't want to wait the community to use these new features.

However, getting a new minor version in every 2-weeks may make the community hard to follow the new features & changes and upgrade to the new version every time (while we don't introduce important breaking changes in minor versions and `abp update` command handles it in most cases). We know, getting behind of the latest version doesn't feel us good :)

**So, we are asking to the community to change the 2-weeks release cycle to 3 or 4 weeks or keep the 2-weeks? We want to get feedback. Even a simple "2", "3" or "4" response is pretty appreciated. Thanks a lot :)**

BTW, we will also start to publish RC versions in a short time (see #4691). This will makes the stable releases more stable.

**Comment by StevenOwenNell** [comment]
I vote to keep it at 2-week intervals to keep the momentum going for now - I think we should ask this question at every major release.

**Comment by abierhaus** [comment]
Ich glaube persönlich, dass die meisten sinnvollen Features nun implementiert sind. Neue und geänderte Features sollten stabiler laufen und besser dokumentiert werden. Ich denke viele neue Features sind für die meisten Nice-To-Have (for example Chat). 

Daher möchte ich einen **4** Wochen Rhytmus vorschlagen. Dieses reduziert den Aufwand für Updates erheblich, von 26  Update zyklen zu 12 pro Jahr.

**Comment by abierhaus** [comment]
I personally believe that most useful features are now implemented. New and changed features should run more stable and be better documented. I think many new features are for most of the Nice-To-Have (for example Chat). 

Therefore I would like to suggest a **4** weeks rhythm. This will reduce the time needed for updates significantly, from 26 update cycles to 12 per year.

**Comment by kgamalseif** [comment]
**4** weeks is good :).

**Comment by wakuflair** [comment]
4 is better

**Comment by antosubash** [comment]
4-weeks is better.

**Comment by xewn** [comment]
2 weeks

**Comment by xewn** [comment]
A release frequency of two weeks is a more appropriate pace. Users who are concerned about instability can choose multiple versions themselves before updating and upgrading.

**Comment by jimmidier** [comment]
2 weeks. Slowing down the pace doesn't change the objective fact of whether the framework needs enhancement or not, and I think there's still space left for new features and enhancement. Also I agree on reconsidering this at every major release.

**Comment by gdlcf88** [comment]
4-weeks is nice. Modules managed by open source communities are hard to be upgraded frequently.

**Comment by JadynWong** [comment]
4 weeks

**Comment by michaelsudnik** [comment]
2 weeks is nice, especially with associated release notes and blog posts, as it gives usets visibility of the potentially high impact new features being introduced and leaves it up to the users if they want to update. 

However, if you think that future development will include less whiz bang features, then 4 weeks might be more appropriate.

**Comment by seanalford** [comment]
I vote for 2 weeks.  Keep the momentum going!  If someone prefers a 4 week cycle they can simply pull the updates every 4 weeks.

**Comment by gmalulu** [comment]
2 weeks

**Comment by geffzhang** [comment]
2 weeks is nice ，Lean Agile

**Comment by zhchlmm** [comment]
I vote for 2 weeks.

**Comment by mifercagri** [comment]
4

**Comment by XuJin186** [comment]
2

**Comment by HiteshAnshani** [comment]
4 weeks but with more good documentation of **each module** if possible.

**Comment by hikalkan** [comment]
Thank you for all the responses and the great attention.
**Please continue to share your responses, ideas and comments.**
We will evaluate all :)

**Comment by buaziz** [comment]
3 weeks release an RC
finalize and release stable on 4th week.

**Comment by hikalkan** [comment]
What about that (close to the response of @buaziz):

1. week: -
2. week: 3.1.0 RC
3. week - (rc2, rc3... if necessary)
4. week: 3.1.0 Final
5. week - (3.1.1, 3.1.2... if necessary)
6. week: 3.2.0 RC
7. week - (rc2, rc3... if necessary)
8. week: 3.2.0 Final
...

We release a stable version in every 4-weeks, but we release the RC in the middle of this period. So;

- We have time (2-weeks) to get feedback, test and stabilize the version. In this time we continue to develop features for the next version, as always.
- We have time to complete the documentation for the new features before the stable version.
- Module developers have time to adapt to the new version. We don't add any code to the RC except bugfix. When we release the stable version, they can immediately update and release their modules.


For anyone only interest in the stable versions, they can update once in 4-weeks.

**Comment by hikalkan** [comment]
thanks to @cotur to draw a visual diagram for my comment :)

![image](https://user-images.githubusercontent.com/1210527/87075323-d1a5b600-c228-11ea-8933-7f83a17c118c.png)

**Comment by gdeswardt** [comment]
First of all awesome job everyone involved for creating such a good framework. Love it.

Documentation and more samples and tutorials are key to drive deeper adoption and wider the community support at this point. 

My vote is to release smaller changes more often!

**Comment by ninomartini** [comment]
I want to say I appreciate your team's hard work.  As for me, it does not matter if it is 2 or 4 weeks as long as new features introduced are documented.

**Comment by shivubasavaiah** [comment]
Great job so far. Keep up the momentum with 2 weeks cycle.

**Comment by AppBotTech** [comment]
3 weeks sounds fine. Hope to see Blazor UI soon. Thanks for all your hard work.

**Comment by FrancoisKr** [comment]
2 weeks, thanks guys!

**Comment by SpiraMira** [comment]
@hikalkan - I think your proposed re-baseline to 4 weeks with RC releases is just fine.  In my opinion, revisiting the release schedule is a good remedy as long as it addresses the following:

- better unit testing and integration testing to reduce bugs (out of the box)
- better abp suite testing and integration
- better abp suite angular support (so it keeps up with the mvc side of the house - razor or blazor UI)
- more abp suite crud generator features (including n-to-1 and n-to-n)
- documentation that's always in sync with the releases

as long as this all improves stability ...

Thanks guys

**Comment by dicksonkimeu** [comment]
2 Weeks

**Comment by lanpin** [comment]
2

**Comment by Qboat** [comment]
2 weeks

**Comment by hikalkan** [comment]
Thank you all for your great interest in this topic. 

I see that 2 and 4 are almost equals.

We will try 4-weeks development/release cycle for 3.x versions to see if it is effective for the development and stability. If it not works as we expect, we may always change it in the future.

In addition to the [comments I've done before](https://github.com/abpframework/abp/issues/4692#issuecomment-656251469), this new stable/rc release cycle may have some other additional benefits. For example (to be honest), we see that completing a medium to large feature takes more than 2 weeks (more than one milestone currently). It sometimes leads us to select smaller features to add to the milestone. 4-weeks development cycle may motive us to work on bigger features and we don't worry to not add a significant feature for the next release.

We see.. I hope this change brings more complete, stable and documented features :)

Thank you all for commenting on this issue.

**Comment by zhishile** [comment]
2 weeks
